docs/docs/user-guide/background/megatron_datasets.md

```diff
@@ -56,6 +56,42 @@ for sample in MultiEpochDatasetResampler(dataset, num_epochs=3, shuffle=True):
     [PRNGResampleDataset][bionemo.core.data.resamples.PRNGResampleDataset] offers a simple solution for shuffling a
     dataset with replacement in O(1) memory.
 
+## Training Resumption
+To ensure identical behavior with and without job interruption, BioNeMo provides [MegatronDataModule][bionemo.llm.data.datamodule.MegatronDataModule] to save and load state dict for training resumption, and provides [WrappedDataLoader][nemo.lightning.data.WrappedDataLoader] to add a `mode` attribute to [DataLoader][torch.utils.data.DataLoader].
+
+```python
+class MyDataModule(MegatronDataModule):
+    def __init__(self, *args, **kwargs):
+        super().__init__()
+        ...
+
+    def train_dataloader(self):
+        return WrappedDataLoader(
+            ...,  # any other arguments for DataLoader
+            mode="train",
+        )
+
+    def val_dataloader(self):
+        return WrappedDataLoader(
+            ...,  # any other arguments for DataLoader
+            mode="validation",
+        )
+
+    def test_dataloader(self):
+        return WrappedDataLoader(
+            ...,  # any other arguments for DataLoader
+            mode="test",
+        )
+```
+
+!!! note "MegatronDataModule"
+
+    Users will see non-overlapping training curve if their datamodule is not inheritting from `MegatronDataModule`, unless similar logics are handled by the users. `MegatronDataModule` ensures that training resumes with the correct sample index instead of restarting from 0 everytime.
+
+!!! note "WrappedDataLoader"
+
+    The WrappedDataLoader class is a wrapper around the PyTorch DataLoader class that adds the `mode` attribute to the dataloader. The dataloader will resume from the last sample index only when mode is 'train'. `val_dataloader` and `test_dataloader` are unaffected.
```

Comment:
Also note that `train` is the default mode. So if you forget to return a wrapped dataloader than your val/test classes will change every time you restart since their consumed_samples will be set to a non-zero value.

Reply:
Addressed.